app/src/main/java/de/westnordost/streetcomplete/data/osm/edits/upload/ElementEditUploader.kt

```diff
@@ -1,39 +1,65 @@
 
     /** Apply the given change to the given element and upload it
      *
      *  @throws ConflictException if element has been changed server-side in an incompatible way
      *  */
     fun upload(edit: ElementEdit, idProvider: ElementIdProvider): MapDataUpdates {
-        val element = edit.fetchElement()
-
-        val mapDataChanges = edit.action.createUpdates(edit.originalElement, element, mapDataApi, idProvider)
+        val remoteChanges by lazy { edit.action.createUpdates(edit.originalElement, edit.fetchElement(mapDataApi), mapDataApi, idProvider) }
+        val localChanges by lazy { edit.action.createUpdates(edit.originalElement, edit.fetchElement(mapDataController), mapDataController, idProvider) }
+        val useRemoteChanges = edit.action::class in EDIT_ACTIONS_NOT_ALLOWED_TO_USE_LOCAL_CHANGES
 
         return try {
-            val changesetId = changesetManager.getOrCreateChangeset(edit.type, edit.source)
-            mapDataApi.uploadChanges(changesetId, mapDataChanges)
+            if (useRemoteChanges)
+                uploadChanges(edit, remoteChanges, false)
+            else
+                uploadChanges(edit, localChanges, false)
         } catch (e: ConflictException) {
-            val changesetId = changesetManager.createChangeset(edit.type, edit.source)
-            mapDataApi.uploadChanges(changesetId, mapDataChanges)
+            // either changeset was closed, or element modified, or local element was cleaned from db
+            if (useRemoteChanges) {
+                // probably changeset closed
+                uploadChanges(edit, remoteChanges, true)
+            } else {
+                // anything of the 3 may be the problem, try again with remote changes
+                try {
+                    uploadChanges(edit, remoteChanges, false)
+                } catch (e: ConflictException) {
+                    // probably changeset closed
+                    uploadChanges(edit, remoteChanges, true)
+                }
+            }
         }
     }
 
-    private fun ElementEdit.fetchElement() = when (elementType) {
-        ElementType.NODE     -> mapDataApi.getNode(elementId)
-        ElementType.WAY      -> mapDataApi.getWay(elementId)
-        ElementType.RELATION -> mapDataApi.getRelation(elementId)
+    private fun uploadChanges(edit: ElementEdit, mapDataChanges: MapDataChanges, newChangeset: Boolean): MapDataUpdates {
+        val changesetId = if (newChangeset) changesetManager.createChangeset(edit.type, edit.source)
+            else changesetManager.getOrCreateChangeset(edit.type, edit.source)
+        return mapDataApi.uploadChanges(changesetId, mapDataChanges)
+    }
+
+    private fun ElementEdit.fetchElement(mapDataRepository: MapDataRepository) = when (elementType) {
+        ElementType.NODE     -> mapDataRepository.getNode(elementId)
+        ElementType.WAY      -> mapDataRepository.getWay(elementId)
+        ElementType.RELATION -> mapDataRepository.getRelation(elementId)
     }
```

Comment:
This is actually a weird choice for an extension function. Maybe I wrote this when I was new to extension functions. Now, I would write it like this:

```suggestion
    private fun MapDataRepository.get(elementType: ElementType, elementId: Long) = when (elementType) {
        ElementType.NODE     -> getNode(elementId)
        ElementType.WAY      -> getWay(elementId)
        ElementType.RELATION -> getRelation(elementId)
    }
```

Would you agree that this is better? (Or meinetwegen ElementEdit as parameter)

Reply:
Right, this looks better.
The way I had is was (probably, can't remember) just because it's least changes from original `ElementEdit.fetchElement`.

Reply:
I renamed `get` to `fetch`, so conflicts with `MapDataController.get` are avoided